src/atomic.rs

```diff
@@ -0,0 +1,860 @@
+    /// use crossbeam_epoch::{self as epoch, Atomic, Ptr};
+    /// use std::sync::atomic::Ordering::SeqCst;
+    ///
+    /// let a = Atomic::new(1234);
+    /// a.store(Ptr::null(), SeqCst);
+    /// ```
+    pub fn store(&self, new: Ptr<T>, ord: Ordering) {
+        self.data.store(new.data, ord);
+    }
+
+    /// Stores an `Owned` into the atomic pointer.
+    ///
+    /// This method takes an [`Ordering`] argument which describes the memory ordering of this
+    /// operation.
+    ///
+    /// [`Ordering`]: https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_epoch::{self as epoch, Atomic, Owned};
+    /// use std::sync::atomic::Ordering::SeqCst;
+    ///
+    /// let a = Atomic::null();
+    /// a.store_owned(Owned::new(1234), SeqCst);
+    /// ```
+    pub fn store_owned(&self, new: Owned<T>, ord: Ordering) {
+        let data = new.data;
+        mem::forget(new);
+        self.data.store(data, ord);
+    }
+
+    /// Stores a `Ptr` into the atomic pointer, returning the previous `Ptr`.
+    ///
+    /// This method takes an [`Ordering`] argument which describes the memory ordering of this
+    /// operation.
+    ///
+    /// [`Ordering`]: https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_epoch::{self as epoch, Atomic, Owned, Ptr};
+    /// use std::sync::atomic::Ordering::SeqCst;
+    ///
+    /// let a = Atomic::new(1234);
+    /// epoch::pin(|scope| {
+    ///     let p = a.swap(Ptr::null(), SeqCst, scope);
+    /// });
+    /// ```
+    pub fn swap<'scope>(&self, new: Ptr<T>, ord: Ordering, _: &'scope Scope) -> Ptr<'scope, T> {
+        Ptr::from_data(self.data.swap(new.data, ord))
+    }
+
+    /// Stores `new` into the atomic pointer if the current value is the same as `current`.
+    ///
+    /// The return value is a result indicating whether the new pointer was written. On failure the
+    /// actual current value is returned.
+    ///
+    /// This method takes an [`Ordering`] argument which describes the memory ordering of this
+    /// operation.
+    ///
+    /// [`Ordering`]: https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_epoch::{self as epoch, Atomic, Ptr};
+    /// use std::sync::atomic::Ordering::SeqCst;
+    ///
+    /// let a = Atomic::new(1234);
+    ///
+    /// epoch::pin(|scope| {
+    ///     let mut curr = a.load(SeqCst, scope);
+    ///     let res = a.compare_and_swap(curr, Ptr::null(), SeqCst, scope);
+    /// });
+    /// ```
+    pub fn compare_and_swap<'scope>(
+        &self,
+        current: Ptr<T>,
+        new: Ptr<T>,
+        ord: Ordering,
+        _: &'scope Scope,
+    ) -> Result<(), Ptr<'scope, T>> {
+        let fail_ord = strongest_failure_ordering(ord);
+        match self.data.compare_exchange(current.data, new.data, ord, fail_ord) {
```

Comment:
I feel it would be more convenient to just use `self.data.compare_and_swap` instead.  What do you think?